### ISSUE TITLE
null_ could return the remaining stream

### ISSUE THREAD

**Issue by andrewthad** [michaelt/streaming-bytestring#8]
For testing to see whether or not a stream is exhausted, there are currently two options. There is `null`, which evaluates all the effects and discards all elements in the stream. And there is `null_`, which evaluates effects up to the next non-empty chunk:

    null :: Monad m => ByteString m r -> m (Of Bool r)
    null_ :: Monad m => ByteString m r -> m Bool

I need something like this:

    null_' :: Monad m => ByteString m r -> m (Of Bool (ByteString m r))

Which would tell me if the stream is exhausted, but if it's not, give it back to me, after having stripped off and evaluated any necessary monadic effects.

**Comment by andrewthad** [comment]
I guess that my real problem is that the semantics of `null_` seem weird to me. You don't get a stream back from it, so you cannot safely use the original stream anymore since some monadic effect might get evaluated a second time. I'm sure that this is useful in some situations, but it seems like it would be much more useful to have access to the remaining stream.

**Comment by michaelt** [comment]
I definitely considered such a function. I wonder if something like `testNull` or `checkNull` or something is a better name? I think what impeded me was the thought that it should have a type like 

     Monad m => ByteString m r -> m (Compose (Of Bool) (ByteString m) r)

which seemed slightly mad! That way it could be mapped over a succession of bytestreams, not that I can see what use that might be, but it fits with larger principles. But the type you mention could be employed that way with `liftM Compose . testNull` and your type is much more intelligible.

**Comment by michaelt** [comment]
I added what I think is a correct implementation, temporarily called `testNull` see what you think. https://github.com/michaelt/streaming-bytestring/blob/master/Data/ByteString/Streaming.hs#L449

**Comment by andrewthad** [comment]
Yeah, thats exactly what I had in mind. I'm not crazy about the name `testNull`, but I'm having trouble coming up with anything better. Here are some attempts: `nullAnd`, `nullCont`, `nullTest`, `isNull`, `nullThen`, `nullNext`.
